payload-functions.js

```diff
@@ -182,49 +182,88 @@ function prepareDeleteSubnetPayload(params) {
 }
 
 function validateAddSecurityGroupRulesParams(params) {
-  if (params.fromPorts?.length !== params.toPorts?.length) {
-    throw new Error("From Ports and To Ports must be the same length");
+  if (params.ipProtocol !== "ICMP" && params.ipProtocol !== "All") {
+    if (!params.fromPorts?.length || !params.toPorts?.length) {
+      throw new Error(`From Ports and To Ports are required for ${params.ipProtocol} protocol.`);
+    }
+    if (params.fromPorts?.length !== params.toPorts?.length) {
+      throw new Error("From Ports and To Ports must be the same length.");
+    }
+  } else if (params.ipProtocol === "ICMP" && (params.fromPorts?.length || params.toPorts?.length)) {
+    throw new Error("Ports cannot be configured for protocol ICMP, use parameter \"ICMP Type\" instead.");
   }
-  const allowedPortsForProtocolAll = ["*", "-1", "0-65535"];
-  const areFromPortsLegal = params.fromPorts?.every?.((port) => (
-    allowedPortsForProtocolAll.includes(port)
-  ));
-  const areToPortsLegal = params.toPorts?.every?.((port) => (
-    allowedPortsForProtocolAll.includes(port)
-  ));
-  if (params.ipProtocol === "All" && !(areFromPortsLegal && areToPortsLegal)) {
-    throw new Error("Specifying All IP Protocols allows all traffic and cannot be restricted by Port Range. If you intend to allow a specific Port Range, please use TCP or UDP instead.");
+  if (params.ipProtocol === "All") {
+    const allowedPortsForProtocolAll = ["*", "-1", "0-65535"];
+    const areFromPortsLegal = !params.fromPorts?.length || params.fromPorts?.every?.((port) => (
+      allowedPortsForProtocolAll.includes(port)
+    ));
+    const areToPortsLegal = !params.toPorts?.length || params.toPorts?.every?.((port) => (
+      allowedPortsForProtocolAll.includes(port)
+    ));
+    if (!areFromPortsLegal || !areToPortsLegal) {
+      throw new Error("Specifying All IP Protocols allows all traffic and cannot be restricted by Port Range. If you intend to allow a specific Port Range, please use TCP or UDP instead.");
+    }
   }
 }
 
 function prepareAddSecurityGroupRulesPayload(params) {
   validateAddSecurityGroupRulesParams(params);
-  const ipv6Ranges = params.cidrIps6.map((CidrIpv6) => ({
+
+  const ipProtocol = params.ipProtocol === "All" ? "-1" : params.ipProtocol.toLowerCase();
+  const ipv6Ranges = params.cidrIps6?.map((CidrIpv6) => helpers.removeUndefinedAndEmpty({
     CidrIpv6,
     Description: params.description,
   }));
-
-  const ipRanges = params.cidrIps.map((CidrIp) => ({
+  const ipRanges = params.cidrIps?.map((CidrIp) => helpers.removeUndefinedAndEmpty({
     CidrIp,
     Description: params.description,
   }));
 
-  const mappedPorts = params.fromPorts?.map ? params.fromPorts.map((fromPort, index) => ({
-    FromPort: fromPort,
-    ToPort: params.toPorts[index],
-    IpProtocol: params.ipProtocol,
-  })) : [];
+  const ipPermissionsForAllProtocols = [
+    helpers.removeUndefinedAndEmpty({
```

Comment:
Is there a chance that any of these values will be undefined or empty?

Let's validate the parameters before we start processing them, and not create structures that may contain empty values, then there will be no need to call that helper function.